Add the elapsed property

Kentzo opened the pull request:
- Until started it's 0.0
- Before exit it's now - start
- After exit it's exit - start

Make sure that the `remaining` property is not updated after context manager exits.

asvetlov commented:
I believe just `.cancel_at` (or maybe `.expire_at`)  property is enough: *remained* and *elapsed* could be calculate from the property and `loop.time()`.

One property is better than two.
Also as you've mentioned `.remained` and `.elapsed` are not accurate but `.expire_at` is.

Kentzo commented:
@asvetlov `cancel_at` is not enough as it does take into account only expected time when cancel will be raised, not when block actually exits. And it can exit way after due to a non-asyncio blocking operation. My comment regarding inaccuracy is also regarding this use-case.

Kentzo commented:
Rebased on top of master.

Kentzo commented:
To further explain the cases I'm thinking of:

At the beginning:

```
    +--------remaining---------+

    +-------------------------->
started_at                  cancel_at

    +elapsed
```

During:
```
              +---remaining----+

    +---------*---------------->
started_at                  cancel_at

    +-elapsed-+
```

At the end:
```
                               +remaining

    +-------------------------->          *exited_at
started_at                  cancel_at

    +---------------elapsed---------------+
```

Kentzo commented:
@asvetlov What do you think?

Kentzo commented:
Fixed the merge conflicted and added type definitions for the changes.

The test failed due to 

```
python setup.py check -rms
running check
warning: check: Definition list ends without a blank line; unexpected unindent. (line 60)
```

Kentzo commented:
@asvetlov May I ask what use case changed your mind?

asvetlov commented:
I need an API to advance the scheduled timeout, plus I need both *delay* and *absolute time* (`async with timeout(10):` and `async with timeout.at(now+10)` forms.

Additional introspection API can be useful for these cases.

The only thing I'm still not sure about: should it be `elapsed`/`remaining` pair or `started`/`finished` absolute times. Both notations are equal, something like `when` (exposing `self._cancel_at`) is needed anyway.

Let me work on it, improvements should be a part of *4.0* release.

Kentzo commented:
My original use case was the following: I wanted to allocate a specific # of seconds for a job consisting of multiple async tasks. Sequential tasks then received `remaining` as their own timeouts. Contrary to just relying on receiving cancellation, it allowed tasks to better prioritize their work.

The `elapsed` property was used to report the metrics by how much jobs/tasks outlived allocated time.

asvetlov commented:
The same can be achieved by `started` + `finished` properties and `loop.time()`, right?

Kentzo commented:
I think so, yes.

asvetlov commented:
Sorry, at work on #118 I've ripped out this functionality.
The reason is: started/finished or elapsed/remaining can be implemented *on top* of existing functionality.
The `Timeout` class is short but the design is not trivial; I spent several days trying to collect and implement all required behavior. 
Adding two new properties adds own complexity and increases a chance to miss something.

Also, there are styling questions:
1. I don't like Optional finished/remaining type, every code that calls this property should check for None before any operation
2. Naming is a problem. In #118 `expired` is a boolean, `deadline` is a float absolute time. What  type is for `started` / `finished`? It looks like `expired` but provides absolute time. Brrr.
Elapsed/remaining is not better because it returns an *approximation*, the value obsoletes just after retrieving. Absolute times are better but see the text above.

Kentzo commented:
I think the fact that PR remained opened for so long quite shows that demand for that feature was low.

1. It's fairly save to assume that the practical code shouldn't check for `None`, as an attempt to access these properties while they may be `None` shows errors in control flow.

2. I usually use the `_at` suffix for timestamp variables, works nicely

asvetlov commented:
Agree, agree and agree again.

I want to publish async-timeout 4.0 without this feature and wait for feedback.
This is a plan.
